src/server/conn_context.h

```diff
@@ -209,6 +210,12 @@ class ConnectionContext : public facade::ConnectionContext {
   // Reference to a FlowInfo for this connection if from a master to a replica.
   FlowInfo* replication_flow = nullptr;
 
+  ServerState* SS() const {
+    return server_state;
+  }
+
+  ServerState* server_state = nullptr;
```

Comment:
And this is as error prone as it was previously. The issue is not `ServerState::tlocal`, the issue when:

```
// store thread local into a local variable
// now invoke cmd and migrate thread
// use local variable which creates the data race
```

Nobody is stopping the programmer to do the same, that is:

```
// store server_state in local variable
// migrate
// use local variable 
```

Therefore, because of this and the issue I commented above, I replaced the local variables with explicit calls `ServerState::tlocal()` so even if the thread migrates, the call to `tlocal()` will yield the correct thread local variable

Reply:
that's not exact. A compiler can cache `ServerState::tlocal` result because it knows that the returned value comes from thread-local variable.

Reply:
See my last comment. Unfortunately, it won't fly.